src/jit/optimizer.cpp

```diff
@@ -7003,22 +7015,36 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                 // Is the value of the whole tree loop invariant?
                 if (!treeIsInvariant)
                 {
+                    // Here we have a tree that is not loop invariant and we thus cannot hoist
                     treeIsHoistable = false;
                 }
             }
 
-            // Check if we need to set 'm_beforeSideEffect' to false.
-            // If we encounter a tree with a call in it
-            //  or if we see an assignment to global we set it to false.
+            // Next check if we need to set 'm_beforeSideEffect' to false.
             //
-            // If we are already set to false then we can skip these checks
+            // If we have already set it to false then we can skip these checks
             //
             if (m_beforeSideEffect)
             {
+                // Is the value of the whole tree loop invariant?
+                if (!treeIsInvariant)
```

Comment:
I think the right check here is `if (!treeIsHoistable)`. If the tree is invariant and can throw but we don't hoist it, we need to set `m_beforeSideEffect` to `false`.

Reply:
Unfortunately we have trees that switch back and forth from having treeIsHoistable as true to false and back to true.
Only the switch to treeIsInvariant is permanent.

One such tree that is important to be able to hoist is an invariant array read inside a loop:

Because ARR_BOUNDS_CHECK is a void, (it doesn't produce a value) it isn't a CSE candidate,
and anything that isn't a CSE candidate has treeIsHoistable  set to false.
It's parent node is a GT_COMMA which produces a ushort value and will be hoistable.


```
               [000054] ---XG+------                 +--*  COMMA     ushort
               [000047] ---X-+------                 |  +--*  ARR_BOUNDS_CHECK_Rng void  
               [000009] -----+------                 |  |  +--*  LCL_VAR   int    V02 loc0         
               [000046] ---X-+------                 |  |  \--*  ARR_LENGTH int   
               [000008] -----+------                 |  |     \--*  LCL_VAR   ref    V01 arg1         
               [000010] a--XG+------                 |  \--*  IND       ushort
               [000053] -----+------                 |     \--*  ADD       byref 
               [000044] -----+------                 |        +--*  LCL_VAR   ref    V01 arg1         
               [000052] -----+------                 |        \--*  ADD       long  
               [000050] -----+------                 |           +--*  LSH       long  
               [000048] -----+------                 |           |  +--*  CAST      long <- int
               [000045] i----+------                 |           |  |  \--*  LCL_VAR   int    V02 loc0         
               [000049] -----+-N----                 |           |  \--*  CNS_INT   long   1
               [000051] -----+------                 |           \--*  CNS_INT   long   12 Fseq[#FirstElem]
```

Reply:
We may still have a hole here if we encounter a tree that's invariant and can throw an exception but we never hoist it (directly or as part of a parent). I tried to construct an example but ran into another bug where we do an incorrect exception reordering even without a loop: https://github.com/dotnet/coreclr/issues/27027.
I'm ok with this fix as it makes us "more correct" than before but we'll need to revisit this code.